<a href="https://colab.research.google.com/github/HerraKaava/tensorflow/blob/main/tensorflow_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h3 align="left">Workflow</h3>

In this notebook, we're going to cover some of the most fundamental concepts of tensors using TensorFlow

More specifically, we're going to cover:

* Introduction to tensors
* Getting information from tensors
* Manipulating tensors
* Tensors & NumPy
* Using @tf.function (a way to speed up your regular Python functions)
* Using GPUs with TensorFlow (or TPUs)

In [2]:
import tensorflow as tf
import tensorflow_probability as tfp
import numpy as np

In [3]:
print(tf.__version__)

2.15.0


<br>

<h1 align"center"=>Introduction to Tensors</h1>

<h3>Creating tensors with tf.constant()</h3>

In [4]:
scalar = tf.constant(7)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [5]:
# Check the number of dimensions of a tensor

scalar.ndim

0

In [6]:
# Create a vector

vector = tf.constant([10, 10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [7]:
# Check the dimension of our vector

vector.ndim

1

In [8]:
# Create a matrix

matrix = tf.constant([[10, 7],
                      [7, 10]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [9]:
matrix.ndim

2

In [10]:
# Let's create another matrix with flaot16 dtype

matrix2 = tf.constant([[10., 7.],
                       [3., 2.],
                       [8., 9.]],
                      dtype=tf.float16)
matrix2

<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 3.,  2.],
       [ 8.,  9.]], dtype=float16)>

* By default, dtype with tf.constant() is int32 (32-bit precision).
* The higher the number of precision, the more exact these numbers are stored on your computer.
* So float16 < float32 for the same number n.
* This also means that the lower the precision, the less space it takes to store it on your computer.

In [11]:
matrix2.ndim

2

In [12]:
# Let's create a tensor (ndim > 2)

tensor = tf.constant([[[1, 2, 3],
                       [4, 5, 6]],
                     [[7, 8, 9],
                      [10, 11, 12]]])
tensor

<tf.Tensor: shape=(2, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]], dtype=int32)>

In [13]:
tensor.ndim

3

* The number of elements in the shape object correspond to the number of dimensions in a tensor.
* Notice that the dimensions of a tensor are affected by where you place the square brackets.

What we've created so far:

* scalar: a single number
* vector: a number with direction (e.g., wind speed and direction)
* matrix: a 2-dimensional array of numbers
* tensor: an n-dimensional array of numebrs


Notice that

* 0-dimensional tensor is a scalar
* 1-dimensional tensor is a vector
* 2-dimensional tensor is a matrix

<br>

<h3>Creating tensors with tf.Variable()</h3>

In [14]:
changeable_tensor = tf.Variable([10, 7])
unchangeable_tensor = tf.constant([10, 7])

In [15]:
# Let's try to change one of the elements in our changeable tensor

changeable_tensor[0].assign(7)

<tf.Variable 'UnreadVariable' shape=(2,) dtype=int32, numpy=array([7, 7], dtype=int32)>

* Notice that *tf.variable()* objects are **mutable**.
* In Python, an object is considered mutable, if its value can be changed after creating it.
* Notice that the assign() function assigns values in-place.

In [16]:
# Uncomment and run to see the error message

# unchangeable_tensor[0].assign(7)

* AttributeError: 'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'assign'
* This is because *tf.constant()* objects are **immutable**.
* In Python, an object is considered immutable, if its value cannot be changed after it has been created.

**NOTE**

* Rarely in practice you'll need to decide whether to use *tf.constant()* or *tf.variable()* to create tensors, as TensorFlow does this for you.

<br>

<h3>Creating random tensors</h3>

* Random tensors are tensors of some arbitrary size with random numbers.

In [17]:
# Initialize a new random number generator using the seed value 42.
# This part does not initialize any tensor values.
# Here we only set up a generator that can be used to create random values

rng = tf.random.Generator.from_seed(42)

In [18]:
# Assign random data from a standard normal distribution into the rng object

rng = rng.normal(shape=(3,2))

In [19]:
rng

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [20]:
# Shuffle the order of elements in a tensor

arr = tf.constant([[10,7],
                   [3,4],
                   [2,5]])

tf.random.shuffle(arr)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]], dtype=int32)>

* Note that random.shuffle() shuffles the elements along its first dimension.
* Since the first dimension stands for rows (3), only the rows gets shuffled, and the order of the elements in the rows stays the same.

If you want to get the same results for reproducibility, you'll have to set a global random seed.

In [21]:
# Global random seed

tf.random.set_seed(42)
tf.random.shuffle(arr)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 3,  4],
       [ 2,  5],
       [10,  7]], dtype=int32)>

* This produces the same pseudo random numbers every time the code is ran.

In [22]:
tf.random.shuffle(arr, seed=42)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

* This does not produce the same pseudo random numbers, because specifying the seed number inside a function is an operation level seed (not global).
* If both the global and the operation seed are set, both seeds are used in conjunction to determine the random sequence.
* This means that specifying both global and operation level seeds will produce different pseudo random numbers than just specifying a global random seed.
* [tf.random.set_seed](https://www.tensorflow.org/api_docs/python/tf/random/set_seed)

<br>

<h3>Creating tensors from NumPy arrays</h3>

In [23]:
# Create a tensor of ones

tf.ones([4, 4])

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]], dtype=float32)>

In [24]:
# Create a tensor of zeros

tf.zeros(shape=(4,4))

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

**Turning NumPy arrays into tensors**

* The main difference between NumPy arrays and TensorFlow tensors is that tensors can be run on a GPU (much faster for numerical computing).

In [25]:
numpy_arr = np.arange(1, 25, dtype=np.int32)

In [26]:
tf.constant(numpy_arr)

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

* As seen above, NumPy arrays can be directly converted into tf tensors.
* We can also modify the shape of the converted numpy array as long as the size matches the new dimensions.
* For example, since *numpy_arr* has size 24, we can change its shape into (2,3,4), because $2*3*4 = 24.$

In [27]:
tf.constant(numpy_arr, shape=(2,3,4))

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]], dtype=int32)>

In [28]:
tf.constant(numpy_arr, shape=(3,8), dtype=np.int32)

<tf.Tensor: shape=(3, 8), dtype=int32, numpy=
array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [17, 18, 19, 20, 21, 22, 23, 24]], dtype=int32)>

<br>

<h3>Getting information from your tensors</h3>

Some important tensor attributes include

* shape
  * the length (number of elements) of each of the dimensions of a tensor.

* rank
  * the number of tensor dimensions.

* axis or dimension
  * a particular dimension of a tensor.

* size
  * the total number of elements in the tensor.

In [29]:
# Create a rank 4 tensor

t4 = tf.zeros(shape=(2,3,4,5))
t4

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [30]:
t4.ndim

4

In [31]:
# Play with this indexing, and you'll understand how to slice tensors.

t4[0]

<tf.Tensor: shape=(3, 4, 5), dtype=float32, numpy=
array([[[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]]], dtype=float32)>

In [32]:
tf.size(t4)

<tf.Tensor: shape=(), dtype=int32, numpy=120>

In [33]:
# To obtain just the size itself, use the numpy() attribute.

tf.size(t4).numpy()

120

In [34]:
2*3*4*5

120

In [35]:
t4[0, 0, :, :]

<tf.Tensor: shape=(4, 5), dtype=float32, numpy=
array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]], dtype=float32)>

<br>

<h3>Indexing tensors</h3>

In [36]:
t4.shape

TensorShape([2, 3, 4, 5])

In [37]:
# Get the first 2 elements of each dimension of the t4 (rank 4) tensor

t4[:2, :2, :2, :2]


<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [38]:
# Get the first element from each dimension from each index except the last one

t4[:1, :1, :1]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

* Notice that if we do not specify what to get from the last dimension, all of it gets returned.

In [39]:
# This returns the same tensor as the one above

t4[:1, :1, :1, :]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [40]:
# Create a rank 2 tensor (2 dimensions)

t2 = tf.constant([[10, 7],
                  [3, 4]])
print(t2)
print(t2.ndim)

tf.Tensor(
[[10  7]
 [ 3  4]], shape=(2, 2), dtype=int32)
2


In [41]:
# Get the last item of each row of our rank 2 tensor (i.e., get the last column)

t2[:, 1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4], dtype=int32)>

In [42]:
# Add a new dimension to our rank 2 tensor

t2[..., tf.newaxis]

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [43]:
# This will also produce a new axis, but in a different place

t2[:, tf.newaxis]

<tf.Tensor: shape=(2, 1, 2), dtype=int32, numpy=
array([[[10,  7]],

       [[ 3,  4]]], dtype=int32)>

* Notice that you need the three dots (...) in order to get a new dimension as the last dimension.

In [44]:
# Alternative to [..., tf.newaxis]

tf.expand_dims(t2, axis=-1)

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

* axis=-1 stands for the last axis (i.e., the new axis should be placed as the last axis).

In [45]:
# We could also place the new axis as the first axis

tf.expand_dims(t2, axis=0)

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]], dtype=int32)>

* The only difference between these two last lines of code (axis=-1 vs. axis=0) is that the numbers are stored differently.

<br>

<h3>Manipulating tensors (basic tensor operations)</h3>

**Basic operations** (element-wise operations)

*  +, -, *, /

In [46]:
# You can add values to a tensor

t = tf.constant([[10, 7],
                 [5, 6]])
t+5

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[15, 12],
       [10, 11]], dtype=int32)>

* Addition to a tensor adds the value to every element of the tensor.

In [47]:
# Multiplication

t * 2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 14],
       [10, 12]], dtype=int32)>

In [48]:
# Subtraction

t - 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 0, -3],
       [-5, -4]], dtype=int32)>

In [49]:
# Division

t / 2

<tf.Tensor: shape=(2, 2), dtype=float64, numpy=
array([[5. , 3.5],
       [2.5, 3. ]])>

In [50]:
# We can use the tensorflow built-in function too

tf.multiply(t, 2)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 14],
       [10, 12]], dtype=int32)>

* If you have to use some sort of operations on your tensor, always choose a function that comes with the tensorflow library (if possible), since those operations run on the GPU and therefore are faster.

<br>

<h3>Matrix multiplication</h3>

In machine learning, matrix multiplication is one of the most common tensor operations. When multiplying two matrices A and B, the matrix B needs to have the same number of rows than the matrix A has columns. That is, if A has the dimensions $m \times n$ and B has the dimensions $k \times p,$ the matrix multiplication AB is defined if and only if $n = k.$ Note that in matrix multiplication the order matters, meaning that $AB \neq BA$ (well sometimes it is, but in most cases, it is not). The resulting matrix AB will have dimensions $m \times p$.

In [51]:
# Matrix multiplication in tensorflow

tf.linalg.matmul(t2, t2)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [52]:
t2 @ t2

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [53]:
np.matmul(t2, t2)

array([[121,  98],
       [ 42,  37]], dtype=int32)

In [54]:
X1 = tf.constant([[4, 5, 2],
                  [9, 4, 1],
                  [5, 6, 7]])

X2 = tf.constant([[7, 8, 9],
                  [5, 5, 3]])

print(f'{X1.shape}\n{X2.shape}')

(3, 3)
(2, 3)


* These matrices cannot be multiplied, because X1 has 3 columns and X2 has only 2 rows ($2 \neq 3$).
* If you try to multiply these as they are, you will get an error.

In [55]:
# X2 needs to have 3 rows for the matrix multiplication, so let's reshape it

tf.matmul(X1, tf.reshape(X2, shape=(3,2)))

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 83,  63],
       [104,  95],
       [124,  91]], dtype=int32)>

* Notice that the new dimensions of the matrix X1X2 are $3 \times 2$.
* This shows that the resulting matrix after a matrix multiplication AB will have the same amount of rows as A, and same amount of columns as B.

In [56]:
# Transpose

tf.transpose(X1)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[4, 9, 5],
       [5, 4, 6],
       [2, 1, 7]], dtype=int32)>

In [57]:
tf.transpose(X2)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[7, 5],
       [8, 5],
       [9, 3]], dtype=int32)>

<br>

<h3>The dot product</h3>

Matrix multiplication is also referred to as the *dot product*.

You can perform matlix multiplication using:

* tf.matmul()
* tf.tensordot()


In [58]:
X1

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[4, 5, 2],
       [9, 4, 1],
       [5, 6, 7]], dtype=int32)>

In [59]:
tf.matmul(X1, tf.reshape(X2, shape=(3,2)))

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 83,  63],
       [104,  95],
       [124,  91]], dtype=int32)>

In [60]:
tf.tensordot(X1, tf.reshape(X2, shape=(3,2)), axes=1)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 83,  63],
       [104,  95],
       [124,  91]], dtype=int32)>

* Notice that with tf.tensordot(), you need to specify along which axis to perform the multiplication.

<br>

<h3>Changing the datatype of tensors</h3>

In [61]:
print(tf.constant([7.4, 5.0]))
print(tf.constant([5, 4]))
print(tf.constant([1.0, 2, 3, 4, 5]))

tf.Tensor([7.4 5. ], shape=(2,), dtype=float32)
tf.Tensor([5 4], shape=(2,), dtype=int32)
tf.Tensor([1. 2. 3. 4. 5.], shape=(5,), dtype=float32)


* As seen above, when including floats into the tensor, the default datatype will be float32.
* On the other hand, if the tensor contains only integers, the default datatype will be int32.
* Note that if the tensor contains even one float, all of the numbers will be converted into floats.

In [62]:
# Let's reduce the precision from float32 --> float16

print(tf.constant([1.0, 2.0], dtype=tf.float16))

tf.Tensor([1. 2.], shape=(2,), dtype=float16)


* [Mixed precision](https://www.tensorflow.org/guide/mixed_precision)
* The link above contains important information on the different datatypes.

<br>

<h3>Aggregating tensors</h3>

* *Aggregation* is the process of combining several numerical values into a single representative value, and an aggregation function performs this operation.
* Examples of aggregation functions include *min(), max(), mean(), std()*, etc...

In [63]:
t = tf.constant([[5.0, 3.0, 8.0],
                 [9.0, 2.0, 3.0],
                 [6.0, 8.0, 4.0]])
print(tf.reduce_min(t, keepdims=False))
print(tf.reduce_max(t, keepdims=False))
print(tf.reduce_mean(t, keepdims=False))
print(tf.reduce_sum(t, keepdims=False))

tf.Tensor(2.0, shape=(), dtype=float32)
tf.Tensor(9.0, shape=(), dtype=float32)
tf.Tensor(5.3333335, shape=(), dtype=float32)
tf.Tensor(48.0, shape=(), dtype=float32)


In [64]:
(5+3+8+9+2+3+6+8+4)

48

In [65]:
(5+3+8+9+2+3+6+8+4) / 9

5.333333333333333

* The rank of the tensor is reduced by 1 for each dimension that is reduced.
* Note that if no *axis* argument is specified, tf.reduce_SomeAggFunc() reduces the tensor to a scalar (i.e., rank 0), because it computes the mean over all elements in the tensor.

In [66]:
print(tf.math.reduce_variance(t))
print(np.var(t.numpy()))
print()
print(tf.math.reduce_std(t))
print(np.std(t.numpy()))

tf.Tensor(5.7777777, shape=(), dtype=float32)
5.7777777

tf.Tensor(2.4037008, shape=(), dtype=float32)
2.4037008


<br>

<h3>Find the positional minimum and maximum of a tensor (argmin, argmax)</h3>

In [67]:
# Create a new tensor for finding positional min and max (argmin, argmax)

tf.random.set_seed(42)
t = tf.random.uniform(shape=(50,))

# Returns the index with the smallest value across axes of a tensor.
tf.argmin(t).numpy()

16

In [68]:
t[16]

<tf.Tensor: shape=(), dtype=float32, numpy=0.009463668>

* The smallest value in the tensor t is 0.009463668, which is located at the index position 16.

In [69]:
tf.argmax(t)

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [70]:
t[42]

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

* The largest value in the tensor t is 0.9671384, which is located at the index position 42.
* Notice also that when incuding the numpy() function, the output is much nicer, since it returns only the value (see the output of argmin().numpy() above).

In [71]:
# Check for equality

t[tf.argmax(t)] == tf.reduce_max(t)

<tf.Tensor: shape=(), dtype=bool, numpy=True>

<br>

<h3>Squeezing a tensor (removing all single dimensions)</h3>

In [72]:
tf.random.set_seed(42)
t = tf.constant(tf.random.uniform(shape=(50,)), shape=(1,1,1,1,50))

In [73]:
t.shape

TensorShape([1, 1, 1, 1, 50])

In [74]:
tf.squeeze(t).shape

TensorShape([50])

*  squeeze() removes dimensions of size 1 from the shape of a tensor.

<br>

<h3>One-hot encoding tensors</h3>

In [75]:
# Create a list of indices (the categories of some categorical variable)

l_ind = [0, 1, 2, 3]

# one-hot encode our list of indices.
# The depth parameter specifies how many dummy variables you want
# to create from your categorical variable.

tf.one_hot(l_ind, depth=3)

<tf.Tensor: shape=(4, 3), dtype=float32, numpy=
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 0.]], dtype=float32)>

One-hot encoding represents each category with a unique binary vector where only one element is "1" (hot) and the others are "0". For instance, with 4 categories (let's call them A, B, C, and D), the encoding would look like this:

* Category A: [1, 0, 0, 0]
* Category B: [0, 1, 0, 0]
* Category C: [0, 0, 1, 0]
* Category D: [0, 0, 0, 1]

In some contexts, such as when avoiding *multicollinearity* in regression models, it's common to drop one dummy variable. In such cases, if your categorical variable has $n$ categories, you would use $\, n-1\, $ dummy variables. This is because the fourth category can be inferred when all $\, n-1 \,$ dummy variables are 0. This approach ensures that there is no perfect multicollinearity among the dummy variables. Continuing with the example above:

* Category A: [1, 0, 0]
* Category B: [0, 1, 0]
* Category C: [0, 0, 1]
* Category D: [0, 0, 0] (inferred when all other dummy variables are 0)



<br>

<h3>Some more mathematical operations</h3>

In [76]:
t = tf.range(1,11)

In [77]:
tf.square(t)

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([  1,   4,   9,  16,  25,  36,  49,  64,  81, 100], dtype=int32)>

In [78]:
#tf.math.sqrt(t)

* I first tried to run this with a tensor with dtype  int32, and got the following error:
  * Value for attr 'T' of int32 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
* So in tensorflow, the math.sqrt() function does not allow dtype int32

In [79]:
tf.math.sqrt(tf.cast(t, dtype=tf.float32))

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([1.       , 1.4142135, 1.7320508, 2.       , 2.2360678, 2.4494896,
       2.6457512, 2.828427 , 3.       , 3.1622777], dtype=float32)>

* With tf.cast(), we can change the dtype of a tensor.

In [80]:
# Same applies for math.log(), it does not allow int32

tf.math.log(tf.cast(t, dtype=tf.float32))

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 , 1.7917595,
       1.9459102, 2.0794415, 2.1972246, 2.3025851], dtype=float32)>

<br>

<h3>Tensors and NumPy</h3>

**NOTE**: One of the main differences between TensorFlow tensor and NumPy array is that a TensorFlow tensor can be run on a GPU or TPU (for faster numerical processing).

In [81]:
# Create a tensor directly from a numpy array

t = tf.constant(np.array([3., 7., 10.]))
t

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [82]:
# Convert tensor back to numpy array

np.array(t), type(np.array(t))

(array([ 3.,  7., 10.]), numpy.ndarray)

In [83]:
# Tensors (usually) have numpy() method, which converts tensors to numpy arrays

t.numpy()

array([ 3.,  7., 10.])

<br>

<h3>Finding access to GPUs</h3>

* To access a GPU or TPU, navigate to Runtime $\, \boldsymbol{\rightarrow} \,$ Change runtime type

In [84]:
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [88]:
# Check what type of GPU we have access to

!nvidia-smi

Fri May 24 10:27:59 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.05             Driver Version: 535.104.05   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   63C    P0              29W /  70W |    107MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+--

* It seems that we have access to Tesla T4 GPU with more than 16GB of memory.

In [92]:
# 1 MiB is 0.001048576 gigabytes, so our exact memory capacity is

print(f'{np.round(15360 * 0.001048576, 2)}GB')

16.11GB


**NOTE:** If you have access to a CUDA-enabled GPU (we have CUDA Version 12.2), TensorFlow will automatically use it whenever possible.